src/transformers/modeling_utils.py

```diff
@@ -1569,6 +1587,65 @@ def retrieve_modules_from_names(self, names, add_prefix=False, remove_prefix=Fal
 
         return retrieved_modules
 
+    @classmethod
+    def _low_cpu_mem_usage_load(cls, model, loaded_state_dict_keys, resolved_archive_file):
+
+        # dematerialize param storage for keys that are going to be replaced by state_dict, by
+        # putting those on the meta device
+        for k in loaded_state_dict_keys:
+            split_name = k.split(".")
+            m = model
+            matched = False
+            while len(split_name) > 1:
+                if hasattr(m, split_name[0]):
+                    matched = True
+                    m = getattr(m, split_name[0])
+                    del split_name[0]
+                else:
+                    matched = False
+                    break
+            if matched:
+                # selectively switch to the meta device only those params/buffers that will
+                # be next replaced from state_dict. This a complex way to do p.to_("meta")
+                # since we have no in-place to_ for tensors.
+                to_meta_objs = []
+                for n, p in m.named_parameters():
+                    if n == split_name[0]:
+                        to_meta_objs.append(p.data_ptr())
+
+                def selective_to_meta(p):
+                    if p.data_ptr() in to_meta_objs:
+                        return p.to("meta")
+                    else:
+                        return p
+
+                m._apply(selective_to_meta)
+
+        # only now can load state_dict
+        state_dict = torch.load(resolved_archive_file, map_location="cpu")
+
+        # materialize float sd keys one by one on CPU
+        for k in loaded_state_dict_keys:
+            split_name = k.split(".")
+            m = model
+            matched = False
+            while len(split_name) > 1:
+                if hasattr(m, split_name[0]):
+                    matched = True
+                    m = getattr(m, split_name[0])
+                    del split_name[0]
+                else:
+                    matched = False
+                    break
+            if matched:
+                persistent_buffers = {k: v for k, v in m._buffers.items() if k not in m._non_persistent_buffers_set}
+                if split_name[0] in persistent_buffers:
```

Comment:
Why do you need this distinction?
Could this work:
```python
new_val = state_dict[k]
if isinstance(getattr(m, split_name[0]), torch.nn.Parameter):
    new_val = torch.nn.Parameter(new_val)
setattr(m, split_name[0], new_val)
```

Reply:
I had a boolean persistent buffer and it throws an exception on:
```
setattr(m, split_name[0], new_val)
```

Reply:
Ho, that's weird.
That seems to work fine when I test:
```python
import torch

class MyMod(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer("b", torch.ones(2, dtype=torch.bool))

m = MyMod()

sd = m.state_dict()
print(sd)

setattr(m, "b", torch.zeros(2,dtype=torch.bool))

sd = m.state_dict()
print(sd)
```
just gives
```
OrderedDict([('b', tensor([True, True]))])
OrderedDict([('b', tensor([False, False]))])
```

Reply:
ah, you're correct! the error I was getting just for `torch.nn.Parameter`, so your suggestion simplifies things. Thank you!

Please have a look at the latest version and if you have other suggestions for improvement I'd be happy to incorporate those.